lang/compiler.go

```diff
@@ -344,7 +349,8 @@ type AstProgram struct {
 	Now time.Time
 }
 
-func (p *AstProgram) getSpec(ctx context.Context, runtime flux.Runtime, alloc *memory.Allocator) (*flux.Spec, values.Scope, error) {
+// Prepare the AstProgram for semantic analysis
+func (p *AstProgram) Normalize() error {
```

Comment:
Does this function need to be publicly exposed?

Reply:
Yeah, I want to expose it. It is used here: https://github.com/influxdata/influxdb/pull/18796/files#diff-41a88fd64faa5296ede9fdeb08529d1fR195

The problem is that I need to walk the semantic graph to look for option references and the graph is encapsulated inside the compile function. So I want to get the semantic graph by myself. The `extern` piece of code needs to be merged together in order to let our analyzer to make sense of the option usages.

Reply:
It looks to me like the primary purpose is to retrieve a merged AST with the externs. Instead of a normalize function, can you add a function to retrieve the merged AST directly?

The main reason is it seems a bit abstract the way that it currently is. When I read the function description, I don't really understand why I would need this. I think it would be more clear if we had a `Get` function that said "returns the normalized ast" rather than a function that changed the state of the compiler.

Reply:
You have to be in the `AstProgram` so you have the context and access to the `extern`. It may need a clearer name. This is used by the compiler itself to mutate its own state anyway, and it is idempotent.

Reply:
That makes sense and I think that's fine. It also makes sense that you would want to reuse the compiler for this as the main way that we access the merged state.

I think the current API is a bit confusing. It seems to imply that it needs to be called to make the AST program work or that it makes some kind of non-idempotent change to the AST. When I hear normalize, I hear it making a change to the contents. I also get concerned with functions that change state because they can be hard to keep track of and it forces `Normalize()` to be reentry safe for being called multiple times.

I think it would be a better API if we created a function that directly returned the AST you want from the compiler or if we added a function to directly do what you want. Since we are adding a function to a program produced by the compiler, we could add an `Analyze` method that merges the AST's and produces an analyzed semantic graph. We would still get the benefit of code reuse but it is my opinion that the API would be more self-documenting.

The problem I find with a normalize function is it isn't clear why you would need to normalize the AST or what that means. And if the purpose of the function is to prepare the AST for semantic analysis we can just have a function that performs semantic analysis.

Reply:
We sometimes need to analyze just an AST without having an `AstProgram` or additional `extern`.
This merge step is very unique to `AstProgram` so I feel still prefer to make it in this way and name it better.

Reply:
I think you might be misunderstanding my concern. I like the function and I like the implementation. I just want the function to return the merged AST so that we are self-documenting how to use the interface. I want to be able to write this sentence.

> If you want to retrieve the final AST from the AstCompiler, you can use this method and it will merge any externs if they exist or it will return the compiled AST.

When you create a function that mutably changes the object and then retrieve the value form the mutable object, it becomes harder to know when you're supposed to call that function. If I'm an external user, how do I know that I need to call the merge function before I get the AST attribute? That's the reason why I would prefer a function that retrieves the merged AST and let the compiler figure out if it needs to merge it or not.